### PR TITLE
Ig 7

### PR DESCRIPTION
**Pull request by igorbacz**

### CODE REVIEW

File: views/forgot-password.ejs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<!DOCTYPE html>
```
Comment (pdybowski) [inline views/forgot-password.ejs:1]: remove

File: views/reset-password.ejs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<!DOCTYPE html>
+<html lang="en">
```
Comment (pdybowski) [inline views/reset-password.ejs:2]: remove

**Review by pdybowski — CHANGES_REQUESTED** [review]
please check Patrycja's PR

File: views/forgot-password.ejs (hunk 1 above)
Comment (igorbacz) [reply views/forgot-password.ejs]: removed

File: views/reset-password.ejs (hunk 2 above)
Comment (igorbacz) [reply views/reset-password.ejs]: removed

File: startup/routes.js [hunk 3]
```diff
@@ -1,23 +1,23 @@
...
-    );
+    res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept, Authorization, X-Auth-Token");
     if (req.method === "OPTIONS") {
       res.header("Access-Control-Allow-Methods", "PUT, POST, PATCH, DELETE");
       return res.status(200).json({});
     }
     next();
   });
 
+  app.use("/forgot-password", passwordReset);
```
Comment (pdybowski) [inline startup/routes.js:19]: this can be places inside user router

File: services/updateUser.service.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    await User.updateOne(
+      { email: body.email },
+      {
+        $set: body,
+      }
+    );
+    return res.status(200).json("updatedUser");
+  } catch (err) {
+    return err;
+  }
```
Comment (pdybowski) [inline services/updateUser.service.js:14]: try catch should be included in controller, not in service

File: services/resetPass.service.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
+const User = require("../models/user.model");
```
Comment (pdybowski) [inline services/resetPass.service.js:1]: content of this file can be placed in user service

**Review by pdybowski — CHANGES_REQUESTED** [review]
please check master branch and see what should be changed in your branch. Moreover please read some notes from me

File: startup/routes.js (hunk 3 above)
Comment (igorbacz) [reply startup/routes.js:19]: done

File: services/updateUser.service.js (hunk 4 above)
Comment (igorbacz) [reply services/updateUser.service.js:14]: done

File: controllers/reset.controller.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
+const { resetService, findService } = require("../services/auth-user.service");
```
Comment (pdybowski) [inline controllers/reset.controller.js:1]: reset.controller should be a part of auth.controller. Same for service and routes

File: startup/routes.js [hunk 7]
```diff
@@ -21,6 +19,7 @@ module.exports = function (app) {
   });
 
   //place routes here ...
-  app.use("/user", authToken, userMealPlanRouter);
+  app.use("/forgot-password", passwordReset);
```
Comment (pdybowski) [inline startup/routes.js:22]: this should be inside auth route

File: startup/routes.js [hunk 8]
```diff
@@ -1,18 +1,16 @@
 const cors = require("cors");
 const express = require("express");
-const userMealPlanRouter = require("../routers/userMealPlan.router");
+const passwordReset = require("../routers/passwordReset");
+const updateUser = require("../routers/updateUser");
```
Comment (pdybowski) [inline startup/routes.js:4]: move this route to the user router

File: services/updateUser.service.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+const User = require("../models/user.model");
```
Comment (pdybowski) [inline services/updateUser.service.js:1]: move UpdateUser service to user service

File: services/auth-user.service.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
+// put api for user
```
Comment (pdybowski) [inline services/auth-user.service.js:1]: move this file into auth service

File: services/auth-user.service.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+const User = require("../models/user.model");
+const jwt = require("jsonwebtoken");
+const { NotFoundError } = require("../helpers/errorHandlers");
+
+exports.resetService = async (email) => {
```
Comment (pdybowski) [inline services/auth-user.service.js:6]: dont know whats that.

File: services/auth-user.service.js [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    id: user._id,
+  };
+
+  const token = jwt.sign(payload, secret, { expiresIn: "15m" });
+  const link = `http://localhost:3000/forgot-password/${user._id}/${token}`;
+  console.log(link);
+  return link;
+};
+
+exports.findService = async (id, token) => {
```
Comment (pdybowski) [inline services/auth-user.service.js:24]: dont know why do wee need that function

File: services/auth-user.service.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+  const secret = process.env.ACCESS_TOKEN_SECRET + user.password;
+
+  return jwt.verify(token, secret, (err, decoded) => {
+    if (err) throw new UnauthorizedError("You don't have access.");
+
+    return decoded;
+  });
+};
+
+exports.resetPassService = async (id, password) => {
```
Comment (pdybowski) [inline services/auth-user.service.js:37]: just resetPassword. Function itself is not a service

File: services/auth-user.service.js [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+  const user = await User.findOne({ id: id });
+
+  if (id !== user.id) throw new NotFoundError("Invalid id.");
```
Comment (pdybowski) [inline services/auth-user.service.js:40]: "User not found" 

Users themselves dont know what id they have

File: services/auth-user.service.js [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+  const secret = process.env.ACCESS_TOKEN_SECRET + user.password;
+
+  return jwt.verify(token, secret, async (err, decoded) => {
+    if (err) throw new UnauthorizedError("You don't have access.");
```
Comment (pdybowski) [inline services/auth-user.service.js:45]: "Access denied"

File: services/auth-user.service.js [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+  const user = await User.findOne({ id: id });
+  if (id !== user.id) throw new NotFoundError("User not found.");
+
+  const secret = process.env.ACCESS_TOKEN_SECRET + user.password;
+
+  return jwt.verify(token, secret, (err, decoded) => {
+    if (err) throw new UnauthorizedError("You don't have access.");
+
+    return decoded;
+  });
```
Comment (pdybowski) [inline services/auth-user.service.js:34]: we use similar code twice. Please add new function to auth.js and import it from there

File: services/auth-user.service.js [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+const jwt = require("jsonwebtoken");
+const { NotFoundError } = require("../helpers/errorHandlers");
+
+exports.resetService = async (email) => {
+  const user = await User.findOne({ email: email });
+  if (!user) {
+    throw new NotFoundError("User not found.");
+  }
+  const secret = process.env.ACCESS_TOKEN_SECRET + user.password;
+  console.log(secret);
```
Comment (pdybowski) [inline services/auth-user.service.js:12]: remove console.log

File: services/auth-user.service.js [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+const User = require("../models/user.model");
+const jwt = require("jsonwebtoken");
+const { NotFoundError } = require("../helpers/errorHandlers");
+
+exports.resetService = async (email) => {
+  const user = await User.findOne({ email: email });
+  if (!user) {
+    throw new NotFoundError("User not found.");
+  }
+  const secret = process.env.ACCESS_TOKEN_SECRET + user.password;
```
Comment (pdybowski) [inline services/auth-user.service.js:11]: do we need this + user.password here?

File: services/auth-user.service.js [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+  const link = `http://localhost:3000/forgot-password/${user._id}/${token}`;
+  console.log(link);
+  return link;
+};
+
+exports.findService = async (id, token) => {
+  const user = await User.findOne({ id: id });
+  if (id !== user.id) throw new NotFoundError("User not found.");
+
+  const secret = process.env.ACCESS_TOKEN_SECRET + user.password;
```
Comment (pdybowski) [inline services/auth-user.service.js:28]: do we need to include password here?

**Review by pdybowski — CHANGES_REQUESTED** [review]
please fix
and watch: [](https://www.youtube.com/watch?v=YSUKhyQvlco&ab_channel=helloroman)

File: services/auth-user.service.js (hunk 17 above)
Comment (igorbacz) [reply services/auth-user.service.js:12]: done

File: services/auth-user.service.js (hunk 15 above)
Comment (igorbacz) [reply services/auth-user.service.js:45]: done

File: controllers/user.controller.js [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+    await UserService.resetUserPassword(id, password);
+    res.send(result);
+  } catch (error) {
+    next(error);
+  }
+};
+
+exports.updateUser = async (req, res, next) => {
+  const updatedData = req.body;
+  this.updateUser(req.body);
```
Comment (pdybowski) [inline controllers/user.controller.js:82]: ?

File: controllers/user.controller.js [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+    const newMealPlan = await UserService.createUserMealPlan(req.body);
+    res.status(201).json(newMealPlan);
+  } catch (error) {
+    next(error);
+  }
+};
+
+exports.updateUserMealPlan = async (req, res, next) => {
+  try {
+    await UserService.udpateUserMealPlan(req.params.id, req.body);
```
Comment (pdybowski) [inline controllers/user.controller.js:33]: can we response with updated mealPlan, please

File: controllers/user.controller.js [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+    next(error);
+  }
+};
+
+exports.updateUser = async (req, res, next) => {
+  const updatedData = req.body;
+  this.updateUser(req.body);
+  try {
+    await UserService.updateUser(req.body);
+    res.send(result);
```
Comment (pdybowski) [inline controllers/user.controller.js:85]: you dont return result anywhere

File: index.js [hunk 23]
```diff
@@ -3,20 +3,25 @@ require("dotenv").config();
 const winston = require("winston");
 const express = require("express");
 const app = express();
-
+const path = require("path");
```
Comment (pdybowski) [inline index.js:6]: do we need this?

File: services/user.service.js [hunk 24]
```diff
@@ -1 +1,108 @@
...
+  const mealPlan = await MealPlan.findById({ _id: id });
+  if (!mealPlan) {
+    throw new NotFoundError("Meal plan doesn't exist");
+  }
+  await MealPlan.deleteOne({ _id: id });
+  return true;
+};
+
+exports.createUserMealPlan = async (reqBody) => {
+  const mealPlan = new MealPlan({
```
Comment (pdybowski) [inline services/user.service.js:44]: you can just 

new MealPlan({...body})

File: startup/routes.js [hunk 25]
```diff
@@ -21,6 +19,8 @@ module.exports = function (app) {
   });
 
   //place routes here ...
-  app.use("/user", authToken, userMealPlanRouter);
+  app.use("/user", userMealPlanRouter, userRouter);
```
Comment (pdybowski) [inline startup/routes.js:22]: we should validate if user has permissions for this router so I would suggest to put authToken back here ;)

**Review by pdybowski — CHANGES_REQUESTED** [review]

File: controllers/user.controller.js (hunk 21 above)
Comment (igorbacz) [reply controllers/user.controller.js:33]: it's not my code

File: index.js (hunk 23 above)
Comment (igorbacz) [reply index.js:6]: it's not my code

File: services/user.service.js (hunk 24 above)
Comment (igorbacz) [reply services/user.service.js:44]: it's not my code

File: startup/routes.js (hunk 25 above)
Comment (igorbacz) [reply startup/routes.js:22]: it's not my code

**Review by pdybowski — APPROVED** [review]

**Pull request merged**